### ISSUE TITLE
test failures on arm64

### ISSUE THREAD

**Issue by bremner** [greghendershott/racket-mode#331]
I'm see test failures for racket-tests/repl and racket-tests/run when running on a debian arm64 box (I had a similar report from an Ubuntu user). 

I looks a bit like the output from the two tests is getting return to the second one. I've attached a build log.
[build.log](https://github.com/greghendershott/racket-mode/files/2373216/build.log)

I tried following the protocol example in #327, and after adjusting the parameters to run to match what the current version of racket-tests.el is using, it seemed to work ok:
```text
% nc 127.0.0.1 55557
(2 def "display")
(2 ok kernel)
(3 def "displayln")
(3 ok ("/usr/share/racket/collects/racket/private/misc.rkt" 164 10))
(4 describe "display")
(4 ok "<div><h1>(defined in #%kernel, signature unavailable)</h1><br/></div>")
```

One thing that surprised me is that the racket run.rkt process actually eventually starts a regular repl 
```
% /usr/bin/racket /home/bremner/racket-mode-20180908+0+e8850c6/run.rkt 55557 '(run nil 2048 t medium nil nil)'
Welcome to Racket v7.0.
> (+ 1 1)
2
```

**Comment by greghendershott** [comment]
> One thing that surprised me is that the racket run.rkt process actually eventually starts a regular repl

That REPL is what's connected to Emacs' stdin/stdout like any `comint-mode` process.

The TCP connection is a side-channel for command requests and responses (and very recently, step-debug-break "notifications").

(Mixing command I/O with normal I/O is fragile. I was perpetually fixing bugs re corner cases.)

**Comment by greghendershott** [comment]
I'm having trouble getting my bearings so a few quick questions:

- The source being used here -- what commit is this?  Is it commit 2b1c7d4 what's currently tip of `master`?

- Is this only happening for arm64 but OK on other archs/builds?

- You said:
    > I tried following the protocol example in #327, and after adjusting the parameters to run to match what the current version of racket-tests.el is using, it seemed to work ok:

    What parameter(s) do you mean? The port number?  What was it instead of 55557?

Thanks.

**Comment by bremner** [comment]
Greg Hendershott <notifications@github.com> writes:

> I'm having trouble getting my bearings so a few quick questions:
>
> - The source being used here -- what commit is this?  Is it commit 2b1c7d4 what's currently tip of `master`?

Yes.

>
> - Is this only happening for arm64 but OK on other archs/builds?
>

I've only tested amd64 (where the test pass) and arm64 (where they
don't). Because it's an architecture independent package, by default it
only gets built on amd64 on Debian. I believe Ubuntu is running the
tests in arm64 by default.

> - You said:
>     > I tried following the protocol example in #327, and after adjusting the parameters to run to match what the current version of racket-tests.el is using, it seemed to work ok:
>
>     What parameter(s) do you mean? The port number?  What was it instead of 55557?

The parameters I pass are "55557 '(run nil 2048 t medium nil nil)'"
(shown in the last text box of the original message). These are gathered
from snooping on the running test; they happen to be different than the
demo you gave in #327.

**Comment by LocutusOfBorg** [comment]
In Ubuntu we run test on amd64, arm64 armhf i386 ppc64el and s390x, and it fails only on arm64.

**Comment by greghendershott** [comment]
@bremner Thanks. I misunderstood that there was a discrepancy with the command-line parameters. Now I understand that you were simply trying a manual test, using the example from #327, but updating the example to pass `nil` for a value added after that example (to pass in a list of files that are candidates for being instrumented for the step-debugger, which is new).

---

When you tried your manual test on arm64, using  `racket run.rkt <params>` and  `nc`, about how many seconds did you need to wait until `nc` got a reply, and also until you saw the REPL prompt?

I wonder if this is "just" a timeout that is only happening on the arm64 test environment?

(I put "just" in quotes, because the default timeout is about 15 seconds -- but that's quite a long time to wait. So if that is indeed the issue, it could be "fixed" by increasing the timeout -- but maybe that doesn't actually fix things for real-world arm64 users. What _might_ make sense is to make sure the timeout is stupidly high when running in a CI or test env.)

**Comment by bremner** [comment]
Greg Hendershott <notifications@github.com> writes:

> @bremner Thanks. I misunderstood that there was a discrepancy with the
> command-line parameters. Now I understand that you were simply trying
> a manual test, using the example from #327, but updating the example
> to pass `nil` for a value added after that example (to pass in a list
> of files that are candidates for being instrumented for the
> step-debugger, which is new).
>
> ---
>
> When you tried your manual test on arm64, using `racket run.rkt
> <params>` and `nc`, about how many seconds did you need to wait until
> `nc` got a reply, and also until you saw the REPL prompt?
>

About 21s, at 99%cpu, so not I/O bound.

Before the repl prompt, nc doesn't see a server on the port.
Once the racket process is up, answers are quick.

> I wonder if this is "just" a timeout that is only happening on the
> arm64 test environment?
>
> (I put "just" in quotes, because the default timeout is about 15
> seconds -- but that's quite a long time to wait.

It's a bit surprising if the Ubuntu armhf test env is faster than the
arm64, but it's possible, and it wouldn't have to be a huge variation.

> So if that is indeed
> the issue, it could be "fixed" by increasing the timeout -- but maybe
> that doesn't actually fix things for real-world arm64 users. What
> _might_ make sense is to make sure the timeout is stupidly high when
> running in a CI or test env.)

Where is the timeout configured? I tried running with TRAVIS_CI=1, but I
suspect that's only control a timeout on the emacs side. The results
look about the same, with more retries to connect.

d

**Comment by greghendershott** [comment]
So two things:

1. `racket-tests.el` is `let`-binding some timeout values higher when `"TRAVIS_CI=1"`: 
  - `racket-command-timeout`: Used for individual synchronous commands, to wait for a response.
  - `racket--cmd-connect-timeout`: Used for individual commands, to wait for the TCP connection to be ready on startup before sending the command request.
**BUT** there's another variable that's _not_ changed for tests:
  - `racket--cmd-connect-attempts`: This is related to the previous, but it's the number of roughly one-second intervals at which the elisp tries to connect to the racket back-end TCP server.
I should have tests `let`-bind that to a higher value, too -- or combine the `--concect-attempts` and `--connect-timeout` vars into one.

2. Startup should be faster if possible for real users. 20 seconds is... really bad.
  - As it happens, I have a `refactor` branch in which I've been working on breaking the commands into separate `.rkt` files that can be `lazy-require`d.  For me on macOS that shaves a couple seconds off the startup time, from say 5 seconds to 3 seconds. But, I can't merge this to `master` until my PR to melpa to change the "recipe" is merged.
  - By far the biggest improvement would come from the racket-mode `*.rkt` files being compiled to bytecode. Already on `master` is a `M-x racket-mode-start-faster` command, which does a `raco make` for the user. I have no idea how to trigger such a "post-install" step when the user installs from MELPA. But for the debian package, is that something you could do??

**Comment by greghendershott** [comment]
1. I pushed to the `refactor` branch commit 71bf0c3.

2. I'm curious what you think: Would it be possible for your package to run the `racket-mode-start-faster` command (which will do a `raco make` and make the startup much faster)?  If it's _possible_, is it wise?

**Comment by bremner** [comment]
Greg Hendershott <notifications@github.com> writes:

> 1. I pushed to the `refactor` branch commit 71bf0c3.
>
> 2. I'm curious what you think: Would it be possible for your package
> to run the `racket-mode-start-faster` command (which will do a `raco
> make` and make the startup much faster)?  If it's _possible_, is it
> wise?
>

The good news is that running racket-mode-start-faster manually before
running the tests does make the tests pass for me on the arm64 system I
have access to (amdahl.debian.org).

I guess it would be possible to byte-compile just for the tests, but it
seems like cheating in the sense that users would not see the improved
startup time (and would presumable still see timeout related failures).

We don't really have any existing  helpers for (re)-byte compiling
racket in debian packages. I guess if the racket-bits were a racket
package then raco pkg might work at the system level. It still sounds
like a fair amount of work to get racket upgrades to work right.

d

**Comment by greghendershott** [comment]
Oh OK. Sorry I'm so ignorant how the Debian packaging system works. I thought it might be fine to do such a post install step (for users not just tests). But if not that's fine I'll think of something else. Anyway need to improve the melpa install, too.

Deploying the Racket code as a Racket package would address this. OTOH it won't be just-works for most users anymore and the now multiple packages could get out of sync. Hmm...

**Comment by bremner** [comment]
Greg Hendershott <notifications@github.com> writes:

> Oh OK. Sorry I'm so ignorant how the Debian packaging system works. I
> thought it might be fine to do such a post install step (for users not
> just tests). But if not that's fine I'll think of something
> else. Anyway need to improve the melpa install, too.

In general it's fine to do a post-install step. The problem is if a user
installs racket-mode on racket 6.12, and then upgrades racket to 7.0,
the racket-mode package by default has no way of knowing about that. On
second thought though, since I'm maintaining both racket and racket-mode
in debian, I can probably manage to update both, if that seems like the
best way forward.

>
> Deploying the Racket code as a Racket package would address this. OTOH
>it won't be just-works for most users anymore and the now multiple
>packages could get out of sync. Hmm...

Yes, you'd probably have to start having version numbers on the elisp
code, and use those to keep in sync. Or something fancier like having
the elisp communicate it's version to the racket-server.

**Comment by greghendershott** [comment]
I just merged to `master` the branch that had commit 71bf0c3.

I _think_ this will resolve the test failure, because it increases all the applicable timeouts somewhat even when running tests locally (not even on CI).  Please let me know?

As for non-test, actual usage: The branch also reorganizes the source files, putting each command in its own .rkt file and `lazy-require`-ing each -- so they load/expand only as-needed. That should improve startup time by a couple seconds. That's still not as good as I'd like -- but doing more probably is a new/separate issue?  If you agree, let me know by closing this and I'll open a new one for that.

**Comment by bremner** [comment]
Greg Hendershott <notifications@github.com> writes:

> I just merged to `master` the branch that had commit 71bf0c3.
>
> I _think_ this will resolve the test failure, because it increases all
> the applicable timeouts somewhat even when running tests locally (not
> even on CI).  Please let me know?

Preliminary tests look good. The test suite passes on arm64 for 4f53fc1.
I need to fix a Debian tooling issue before I can test the resulting
packages; I'll close the issue when I manage that.

d

**Comment by bremner** [comment]
The results seem to work, so closing.
